[core] Refactor CheckUsage to take typed Message<> args; add Kokkos_ErrorMessage.hpp

janciesko opened the pull request:
Replace with message-based API: call sites pass Message<Type>(...) objects so error content lives in
Kokkos_ErrorMessage.hpp and CheckUsage only performs the condition check.

- Add Kokkos_ErrorMessage.hpp with message types (CalledBeforeInit,
  CalledAfterFini, InstanceConstruction*, InstanceDestructionAfterFini)
  and Message<T>::get() for formatted strings.
- CheckUsage<...>::check() now takes one or two Message arguments
  instead of building strings internally.
- Update parallel_for, parallel_scan, parallel_reduce and Cuda/Serial
  execution space code to pass the new Message-based arguments.
- Refactor execution space constructor/destructor preconditions to
  use CheckUsage with Message types.

Note: string ops happen when the abort condition has been met thus should not incur overhead in case of correct usage. The compound condition requires two messages since it implements two conditions internally using a message each.

masterleinad commented:
There are tons of failing tests
```
The following tests FAILED:
	  1 - Kokkos_CoreUnitTest_Serial_ViewSupport (Subprocess aborted) Kokkos
	  2 - Kokkos_CoreUnitTest_Serial1 (Subprocess aborted)  Kokkos
	  3 - Kokkos_CoreUnitTest_Serial2 (Subprocess aborted)  Kokkos
	  4 - Kokkos_CoreUnitTest_Default (Subprocess aborted)  Kokkos
	  5 - Kokkos_CoreUnitTest_InitializeFinalize (Failed)   Kokkos
	  7 - Kokkos_CoreUnitTest_KokkosP (Subprocess aborted)  Kokkos
	 11 - Kokkos_ProfilingTestLibraryLoad (Subprocess aborted) Kokkos
	 12 - Kokkos_ProfilingTestLibraryCmdLine (Subprocess aborted) Kokkos
	 14 - Kokkos_IncrementalTest_SERIAL (Subprocess aborted) Kokkos
	 21 - Kokkos_ContainersUnitTest_Serial (Subprocess aborted) Kokkos
	 22 - Kokkos_UnitTest_Sort (Subprocess aborted)         Kokkos
	 23 - Kokkos_UnitTest_Random (Subprocess aborted)       Kokkos
	 24 - Kokkos_AlgorithmsUnitTest_StdSet_A (Subprocess aborted) Kokkos
	 25 - Kokkos_AlgorithmsUnitTest_StdSet_B (Subprocess aborted) Kokkos
	 26 - Kokkos_AlgorithmsUnitTest_StdSet_C (Subprocess aborted) Kokkos
	 27 - Kokkos_AlgorithmsUnitTest_StdSet_D (Subprocess aborted) Kokkos
	 28 - Kokkos_AlgorithmsUnitTest_StdSet_E (Subprocess aborted) Kokkos
	 29 - Kokkos_AlgorithmsUnitTest_StdSet_Team_A (Subprocess aborted) Kokkos
	 30 - Kokkos_AlgorithmsUnitTest_StdSet_Team_B (Subprocess aborted) Kokkos
	 31 - Kokkos_AlgorithmsUnitTest_StdSet_Team_C (Subprocess aborted) Kokkos
	 32 - Kokkos_AlgorithmsUnitTest_StdSet_Team_D (Subprocess aborted) Kokkos
	 33 - Kokkos_AlgorithmsUnitTest_StdSet_Team_E (Subprocess aborted) Kokkos
	 34 - Kokkos_AlgorithmsUnitTest_StdSet_Team_F (Subprocess aborted) Kokkos
	 35 - Kokkos_AlgorithmsUnitTest_StdSet_Team_G (Subprocess aborted) Kokkos
	 36 - Kokkos_AlgorithmsUnitTest_StdSet_Team_H (Subprocess aborted) Kokkos
	 37 - Kokkos_AlgorithmsUnitTest_StdSet_Team_I (Subprocess aborted) Kokkos
	 38 - Kokkos_AlgorithmsUnitTest_StdSet_Team_L (Subprocess aborted) Kokkos
	 39 - Kokkos_AlgorithmsUnitTest_StdSet_Team_M (Subprocess aborted) Kokkos
	 40 - Kokkos_AlgorithmsUnitTest_StdSet_Team_P (Subprocess aborted) Kokkos
	 41 - Kokkos_AlgorithmsUnitTest_StdSet_Team_Q (Subprocess aborted) Kokkos
	 42 - Kokkos_UnitTest_SIMD (Subprocess aborted)         Kokkos
	 43 - Kokkos_Example_VirtualFunctions (Subprocess aborted)
	 44 - Kokkos_Example_RelocatableFunction (Subprocess aborted)
```

science-enthusiast commented on core/src/impl/Kokkos_ErrorMessages.hpp:
```diff
@@ -0,0 +1,130 @@
+  Message(string arg1_, string arg2_, string arg3_ = "no-label")
+      : arg1(arg1_), arg2(arg2_), arg3(arg3_) {}
+
+  string a = "Kokkos ERROR: attempting to call";
```
`"Kokkos ERROR: attempting to call"` should be `"Kokkos ERROR: attempting to call "` here at line 66.

janciesko replied:
Thanks for noticing, updating

science-enthusiast commented on core/src/impl/Kokkos_ErrorMessages.hpp:
```diff
@@ -0,0 +1,130 @@
+// add more toString variants for type specific string conversions
+
+// Enumeration of message types
+enum Variant {
```
We could use scoped enum (`enum class`) instead of just `enum`. There is better type safety that way.

Also, it is better to think of another name instead of `Variant`.

janciesko replied:
`Message::Variant`->`Message::Type`?

science-enthusiast commented on core/src/impl/Kokkos_ErrorMessages.hpp:
```diff
@@ -0,0 +1,130 @@
+  string a = "Kokkos ERROR: attempting to call";
+  string b = "() **before** Kokkos::initialize() was called.";
+  string c = " Concerns ";
+  string d = "  with  ";
```
`"  with  "` should be `" with "`.

science-enthusiast commented on core/src/Cuda/Kokkos_Cuda_Instance.cpp:
I suppose that like here, the other places relevant to #8546 are already in the pipeline for update. Then the PR will progress further.

science-enthusiast commented on core/src/impl/Kokkos_CheckUsage.hpp:
```diff
@@ -55,80 +41,35 @@ struct UsageRequires {
-        func_name, exec_policy, meta_data);
+        msg1);
+    Kokkos::Impl::CheckUsage<
+        Kokkos::Impl::UsageRequires::isNotFinalized>::check(msg2);
```
We need to check `isNotFinalized` before `isInitialized`. That is because of how `Kokkos::finalize()` is implemented in `core/src/impl/Kokkos_Core.cpp`. The function `post_finalize_internal()` sets `g_is_initialized` to `false`.

science-enthusiast commented on core/src/impl/Kokkos_CheckUsage.hpp:
```diff
@@ -8,41 +8,27 @@
-// The types CheckUsage and UsageRequires are for specifying run-time checks
-// of correct usage of Kokkos functions and other constructs.
 template <typename... T>
 struct CheckUsage;
```
The following three specializations of `CheckUsage` have to be enclosed by
`// NOLINTBEGIN(bugprone-exception-escape)` and `// NOLINTEND(bugprone-exception-escape)`.

janciesko commented:
The idea of adding the message types is good but we should probably keep the helper function `check_execution_space_constructor_precondition` just avoid code duplication. I'll rebase and update. @science-enthusiast

janciesko committed:
Refactor CheckUsage to use typed error messages

Replace with message-based API:
call sites pass Message<Type>(...) objects so error content lives in
Kokkos_ErrorMessage.hpp and CheckUsage only performs the condition check.

- Add Kokkos_ErrorMessage.hpp with message types (CalledBeforeInit,
  CalledAfterFini, InstanceConstruction*, InstanceDestructionAfterFini)
  and Message<T>::get() for formatted strings.
- CheckUsage<...>::check() now takes one or two Message arguments
  instead of building strings internally.
- Update parallel_for, parallel_scan, parallel_reduce and Cuda/Serial
  execution space code to pass the new Message-based arguments.
- Refactor execution space constructor/destructor preconditions to
  use CheckUsage with Message types.

Co-authored-by: Hariprasad KANNAN <hkannan@gmail.com>
Signed-off-by: Jan Ciesko <jan.ciesko@gmail.com>

janciesko committed:
Fix insideExecEnv check order so "after finalize" is reported correctly

Signed-off-by: Jan Ciesko <jan.ciesko@gmail.com>

janciesko commented:
@masterleinad Should be ok to review now

janciesko committed:
Consolidate CheckUsage call sites behind helper functions
* Add check_parallel_precondition helper
* Add check_view_allocation_precondition helper

Signed-off-by: Jan Ciesko <jan.ciesko@gmail.com>

janciesko commented:
@science-enthusiast Yes, the PR should consolidate all the strings and conditions scattered around the code base. I have pushed an update.

janciesko commented:
[Abort when loading a profiling library with no callbacks](https://github.com/kokkos/kokkos/pull/8920#top)[Abort when loading a profiling library with no callbacks](https://github.com/kokkos/kokkos/pull/8920#top)

 Could be a `CheckUsage` candidate as well.

ndellingwood commented:
Jenkins OPENACC-NVHPC-CUDA-12.2 [timeout](https://cloud1.cees.ornl.gov/jenkins-ci/job/Kokkos/job/PR-8733/15/stages/?selected-node=186) in Kokkos_CoreUnitTest_Default:

```
...
[OPENACC-NVHPC-CUDA-12.2] 6: [==========] Running 8 tests from 1 test suite.
[OPENACC-NVHPC-CUDA-12.2] 6: [----------] Global test environment set-up.
[OPENACC-NVHPC-CUDA-12.2] 6: [----------] 8 tests from defaultdevicetype
[OPENACC-NVHPC-CUDA-12.2] 6: [ RUN      ] defaultdevicetype.compiler_macros
[OPENACC-NVHPC-CUDA-12.2] 6: [       OK ] defaultdevicetype.compiler_macros (1 ms)
[OPENACC-NVHPC-CUDA-12.2] 6: [ RUN      ] defaultdevicetype.bhalf_operators
[OPENACC-NVHPC-CUDA-12.2]  6/23 Test  #6: Kokkos_CoreUnitTest_Default ................................***Timeout 1500.15 sec
```

Adrien-Tab commented on core/src/Kokkos_Parallel.hpp:
```diff
@@ -140,17 +139,16 @@ inline void parallel_for(const std::string& str, const ExecPolicy& policy,
   /** Enforce correct use **/
-  Impl::CheckUsage<Impl::UsageRequires::insideExecEnv>::check("parallel_for",
-                                                              policy);
+  Impl::check_parallel_precondition("parallel_for", policy);
+
```
Are these precondition checks redundant with the ones performed later in `parallel_for(str, policy, functor)` ? The check seems to run multiple times for a single user call.

science-enthusiast replied:
In #7675 the decision was made to add checks directly in all the function overloads, instead of only in the overload that is ultimately called. The logic was to be as informative as possible to the user, when the abort happens. The price to pay is the cost of the same check performed multiple times.

Adrien-Tab commented on core/src/Kokkos_Parallel.hpp:
```diff
@@ -164,8 +162,7 @@ inline void parallel_for(const std::string& str, const size_t work_count,
   /** Enforce correct use **/
-  Impl::CheckUsage<Impl::UsageRequires::insideExecEnv>::check("parallel_for",
-                                                              work_count);
+  Impl::check_parallel_precondition("parallel_for", work_count);
```
same thing here

science-enthusiast replied:
https://github.com/kokkos/kokkos/pull/8733/changes#r3240893844

Adrien-Tab commented on core/src/Kokkos_Parallel_Reduce.hpp:
```diff
@@ -378,8 +377,7 @@ inline std::enable_if_t<!(Kokkos::is_view<ReturnType>::value ||
   /** Enforce correct use **/
-  Impl::CheckUsage<Impl::UsageRequires::insideExecEnv>::check("parallel_reduce",
-                                                              policy);
+  Impl::check_parallel_precondition("parallel_reduce", policy);
```
same thing here

science-enthusiast replied:
https://github.com/kokkos/kokkos/pull/8733/changes#r3240893844

science-enthusiast commented:
Retest this please

science-enthusiast commented on core/src/impl/Kokkos_CheckUsage.hpp:
```diff
@@ -52,82 +38,80 @@ struct UsageRequires {
-         << " Concerns " << meta_data << " with exec policy "
-         << fetch_policy_name(exec_policy) << ".";
-      Kokkos::abort(ss.str().c_str());
+      Kokkos::abort(msg.get().c_str());
```
Maybe, given that the program is about to abort, the cost of constructing a `std::string` inside `get()` may not be too expensive. One way to avoid that cost could be to have a member function (say, it is called `abort()`) inside all the `Message` specializations and call `Kokkos::abort` inside that function.

masterleinad replied:
Also, see https://github.com/kokkos/kokkos/pull/9148/changes/ef6d4187750f89a769cc747256acca893fe191f5.

science-enthusiast replied:
Yes, I see your point. Avoiding a `std::stringstream` and using `+` operator of `std::string` for string concatenation is sufficient, since we are only dealing with `std::string`s, string literals and pointers to string literals.

It will be good to rebase once the PR containing ef6d4187750f89a769cc747256acca893fe191f5 is merged. Could you tell us which PR it belongs to?

masterleinad replied:
That's part of https://github.com/kokkos/kokkos/pull/9148.

science-enthusiast commented on core/src/impl/Kokkos_CheckUsage.hpp:
```diff
@@ -52,82 +38,80 @@ struct UsageRequires {
+  Impl::CheckUsage<Impl::UsageRequires::insideExecEnv>::check(
+      Impl::Message::Message<
+          Impl::Message::Type::InstanceConstructionBeforeInit>(
+          std::string(msg)),
```
It will be better to avoid these `std::string` casts.
